### PR TITLE
feat: calculate reserved memory based on a gradient proportion

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Recently, we discovered that when the RisingWave compute node is deployed on a machine with a large memory capacity, such as 100GB, the default 30% reserved memory (30GB!) is overly conservative. This leads to a high miss ratio in the executor cache and block cache, resulting in memory wastage. To address this, let's calculate reserved memory based on a gradient proportion.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

- Change the algorithm to calculate the reserve memory size. The reserved memory size is calculated based on the following gradient: 
  - 30% of the first 16GB 
  - Plus 20% of the rest

</details>

### CODE REVIEW

**Comment by hzxa21** [comment]
Relevant discussion: #16130

File: src/compute/src/memory/config.rs [hunk 1]
```diff
@@ -61,7 +62,11 @@ pub fn reserve_memory_bytes(opts: &ComputeNodeOpts) -> (usize, usize) {
     });
 
     // Should have at least `MIN_SYSTEM_RESERVED_MEMORY_MB` for reserved memory.
-    let reserved = std::cmp::max(reserved, MIN_SYSTEM_RESERVED_MEMORY_MB << 20);
+    let mut reserved = std::cmp::max(reserved, MIN_SYSTEM_RESERVED_MEMORY_MB << 20);
+    // Limit the maximum reserved memory size if it is not explicitly specified.
+    if opts.reserved_memory_bytes.is_none() {
+        reserved = std::cmp::min(reserved, MAX_SYSTEM_RESERVED_MEMORY_MB << 20);
```
Comment (StrikeW) [inline src/compute/src/memory/config.rs:68]: Why not limit on line 61?
```
min((opts.total_memory_bytes as f64 * SYSTEM_RESERVED_MEMORY_PROPORTION).ceil() as usize, 
MAX_SYSTEM_RESERVED_MEMORY_MB << 20)
```

File: src/compute/src/memory/config.rs [hunk 2]
```diff
@@ -26,6 +26,7 @@ pub const MIN_COMPUTE_MEMORY_MB: usize = 512;
 /// The memory reserved for system usage (stack and code segment of processes, allocation
 /// overhead, network buffer, etc.) in megabytes.
 pub const MIN_SYSTEM_RESERVED_MEMORY_MB: usize = 512;
+pub const MAX_SYSTEM_RESERVED_MEMORY_MB: usize = 8192;
 
 const SYSTEM_RESERVED_MEMORY_PROPORTION: f64 = 0.3;
```
[the remark this answers is not in the document]
Comment (lmatz) [reply src/compute/src/memory/config.rs:31]: actually
we have a variable for this:
```
opts.reserved_memory_bytes
```

never mind

File: src/compute/src/memory/config.rs [hunk 3]
```diff
@@ -55,17 +57,44 @@ pub fn reserve_memory_bytes(opts: &ComputeNodeOpts) -> (usize, usize) {
...
     (reserved, opts.total_memory_bytes - reserved)
 }
 
+/// Calculate the reserved memory based on the total memory size.
+/// The reserved memory size is calculated based on the following gradient:
+/// - 30% of the first 16GB
+/// - 25% of the next 16GB
+/// - 20% of the next 32GB
+/// - 15% of the rest
+fn gradient_reserve_memory_bytes(total_memory_bytes: usize) -> usize {
```
Comment (StrikeW) [inline src/compute/src/memory/config.rs:77]: What does the `gradient reserve memory proportion` mean? May I know the background for those magic numbers in the `RESERVED_MEMORY_LEVELS` and `RESERVED_MEMORY_PROPORTIONS`?

Comment (chenzl25) [reply src/compute/src/memory/config.rs:77]: Previously, we used 30% as reserved memory which would consume too much memory in a machine with a large memory capacity. We want to reserve less memory for large machines but keep the original proportion for small machines, so a gradient proportion approach is introduced.

Comment (StrikeW) [reply src/compute/src/memory/config.rs:77]: IIUC, for a 64GB memory machine, we will reserve 64 * 0.3 = 19.2GB previously. In this PR, we will reserve 16*0.3+16*0.25+32*0.2 = 15.2GB.

Comment (chenzl25) [reply src/compute/src/memory/config.rs:77]: Yes. The unit test includes a spreadsheet detailing various configurations of reserved memory relative to total memory.

**Review by StrikeW — APPROVED** [review]
LGTM

**Review by lmatz — APPROVED** [review]

**Review by fuyufjh — APPROVED** [review]

**Pull request merged**